elasticdl/pkg/kernelnew/kernel.go

```diff
@@ -0,0 +1,80 @@
+package kernelnew
+
+// #cgo LDFLAGS: -L./capi -lkernel_api -lm
+// #include "capi/kernel_api.h"
+import "C"
+import (
+	"elasticdl.org/elasticdl/pkg/commonnew"
+	"fmt"
+	"unsafe"
+)
+
+// SGD kernel
+func SGD(grad *commonnew.Tensor, param *commonnew.Tensor, lr float32) error {
+	if len(grad.Buffer) != len(param.Buffer) {
+		return fmt.Errorf("grad Value size not equal to param")
+	}
+	gradPtr := (*C.float)(unsafe.Pointer(&grad.Buffer[0]))
+	paramPtr := (*C.float)(unsafe.Pointer(&param.Buffer[0]))
+	length := len(grad.Buffer) / int(commonnew.DtypeSize[grad.Dtype])
+	C.SGD(gradPtr, paramPtr, C.float(lr), C.longlong(length))
+	return nil
+}
+
+// SparseSGD kernel
+func SparseSGD(grad *commonnew.IndexedTensor, param *commonnew.EmbeddingTable, lr float32) error {
```

Comment:
There are three cases:
- apply dense gradient to dense parameter
- apply sparse gradient to dense parameter. The gradient of native keras embedding layer is sparse, but the embedding parameter is dense.
- apply sparse gradient to sparse parameter 

So, we need an extra SGDKernel

Reply:
fixed